In [1]:
!pip install azure-storage-blob
!pip install pyarrow
!pip install psycopg2 sqlalchemy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 394.5/394.5 kB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.4/193.4 kB 20.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.7/41.7 kB 4.3 MB/s eta 0:00:00


In [2]:
import pandas as pd
import numpy as np
import json
import requests
from io import StringIO
from azure.storage.blob import BlobServiceClient, BlobClient, ContainerClient
from math import ceil
import datetime
import calendar
from sqlalchemy import create_engine

In [3]:
# Azure Functions
def azure_upload_blob(connect_str, container_name, blob_name, data):
    blob_service_client = BlobServiceClient.from_connection_string(connect_str)
    blob_client = blob_service_client.get_blob_client(container=container_name, blob=blob_name)
    blob_client.upload_blob(data, overwrite=True)
    print(f"Uploaded to Azure Blob: {blob_name}")

def azure_download_blob(connect_str, container_name, blob_name):
    blob_service_client = BlobServiceClient.from_connection_string(connect_str)
    blob_client = blob_service_client.get_blob_client(container=container_name, blob=blob_name)
    download_stream = blob_client.download_blob()
    return download_stream.readall()


In [4]:
# Specify the path to your JSON configuration file
config_file_path = 'config.json'

# Load the JSON configuration file
with open(config_file_path, 'r') as config_file:
    config = json.load(config_file)

# Print the configuration
#Connection_STRING = config["connectionString"]

CONNECTION_STRING_AZURE_STORAGE = config["connectionString"]
CONTAINER_AZURE = 'motorvehiclecrash'

# Initialize the BlobServiceClient
blob_service_client = BlobServiceClient.from_connection_string(CONNECTION_STRING_AZURE_STORAGE)

# Get the container client
container_client = blob_service_client.get_container_client(CONTAINER_AZURE)

mv_crash_df = pd.DataFrame()


# List all blobs in the specified container
blob_list = container_client.list_blobs()
for blob in blob_list:
    print(blob.name)
    blob_client = container_client.get_blob_client(blob=blob.name)
    blob_data = blob_client.download_blob()
    blob_content = blob_data.readall().decode('utf-8')
    df = pd.read_csv(StringIO(blob_content))
    # Display the head of the DataFrame
    print(df.shape)
    # since I have only one csv, I am doing to do the following instructions
    mv_crash_df = df.copy()

motorvehiclecrash.csv


<ipython-input-4-47f907e295d5>:30: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(StringIO(blob_content))


(2082277, 29)


Great success! The slightly modified ETL script was able to retrieve the csv dataset I had stored in the Azure Cloud. Now I can begin the transformations on the data. I begin first by removing the columns I wont need.

In [5]:
def week_of_month(dt):
    first_day = dt.replace(day=1)
    dom = dt.day
    adjusted_dom = dom + first_day.weekday()
    return int(ceil(adjusted_dom/7.0))

def get_week_of_year(date_str):
    """
    Calculate the ISO week number of the year for a given date.

    Parameters:
    date_str (str): A date string in the format 'YYYY-MM-DD'.

    Returns:
    int: ISO week number of the year.
    """
    # Parse the input string to a datetime object
    date = datetime.strptime(date_str, '%Y-%m-%d')

    # Get the ISO calendar week number
    week_of_year = date.isocalendar()[1]

    return week_of_year

In [6]:
# Check for null values in each column
null_values = mv_crash_df.isnull().sum()

# Display the count of null values in each column
print(null_values)

CRASH DATE                             0
CRASH TIME                             0
BOROUGH                           647881
ZIP CODE                          648132
LATITUDE                          234094
LONGITUDE                         234094
LOCATION                          234094
ON STREET NAME                    442554
CROSS STREET NAME                 787961
OFF STREET NAME                  1732096
NUMBER OF PERSONS INJURED             18
NUMBER OF PERSONS KILLED              31
NUMBER OF PEDESTRIANS INJURED          0
NUMBER OF PEDESTRIANS KILLED           0
NUMBER OF CYCLIST INJURED              0
NUMBER OF CYCLIST KILLED               0
NUMBER OF MOTORIST INJURED             0
NUMBER OF MOTORIST KILLED              0
CONTRIBUTING FACTOR VEHICLE 1       6845
CONTRIBUTING FACTOR VEHICLE 2     323327
CONTRIBUTING FACTOR VEHICLE 3    1933393
CONTRIBUTING FACTOR VEHICLE 4    2048639
CONTRIBUTING FACTOR VEHICLE 5    2073154
COLLISION_ID                           0
VEHICLE TYPE COD

In [7]:
df_cleaned = mv_crash_df.copy()
df_cleaned = mv_crash_df.drop(columns=['CONTRIBUTING FACTOR VEHICLE 3', 'CONTRIBUTING FACTOR VEHICLE 4', 'CONTRIBUTING FACTOR VEHICLE 5', 'VEHICLE TYPE CODE 3', 'VEHICLE TYPE CODE 4', 'VEHICLE TYPE CODE 5', 'OFF STREET NAME'])

# Check for null values in each column
null_values = df_cleaned.isnull().sum()

# Display the count of null values in each column
print(null_values)

df_cleaned.shape

CRASH DATE                            0
CRASH TIME                            0
BOROUGH                          647881
ZIP CODE                         648132
LATITUDE                         234094
LONGITUDE                        234094
LOCATION                         234094
ON STREET NAME                   442554
CROSS STREET NAME                787961
NUMBER OF PERSONS INJURED            18
NUMBER OF PERSONS KILLED             31
NUMBER OF PEDESTRIANS INJURED         0
NUMBER OF PEDESTRIANS KILLED          0
NUMBER OF CYCLIST INJURED             0
NUMBER OF CYCLIST KILLED              0
NUMBER OF MOTORIST INJURED            0
NUMBER OF MOTORIST KILLED             0
CONTRIBUTING FACTOR VEHICLE 1      6845
CONTRIBUTING FACTOR VEHICLE 2    323327
COLLISION_ID                          0
VEHICLE TYPE CODE 1               13808
VEHICLE TYPE CODE 2              398959
dtype: int64


(2082277, 22)

Before starting the cleaning of missing/null values, I had 2,082,277 rows of data.

In [8]:
# Drop rows with any null values and refresh the index
df_cleaned.dropna(inplace=True)
df_cleaned.reset_index(drop=True, inplace=True)

# Check for null values in each column
null_values = df_cleaned.isnull().sum()
print(null_values)

df_cleaned.shape


CRASH DATE                       0
CRASH TIME                       0
BOROUGH                          0
ZIP CODE                         0
LATITUDE                         0
LONGITUDE                        0
LOCATION                         0
ON STREET NAME                   0
CROSS STREET NAME                0
NUMBER OF PERSONS INJURED        0
NUMBER OF PERSONS KILLED         0
NUMBER OF PEDESTRIANS INJURED    0
NUMBER OF PEDESTRIANS KILLED     0
NUMBER OF CYCLIST INJURED        0
NUMBER OF CYCLIST KILLED         0
NUMBER OF MOTORIST INJURED       0
NUMBER OF MOTORIST KILLED        0
CONTRIBUTING FACTOR VEHICLE 1    0
CONTRIBUTING FACTOR VEHICLE 2    0
COLLISION_ID                     0
VEHICLE TYPE CODE 1              0
VEHICLE TYPE CODE 2              0
dtype: int64


(893205, 22)

After I dropped the null values from all the columns, I was left with 893K+ rows of data. I converted the data type of the ZIP CODE column into an integer to utilize it when creating the Location Dimension.

In [9]:
# Display any unique values in 'ZIP CODE' that might not be numeric or are placeholders
print("Unique non-numeric or placeholder values in 'ZIP CODE':", df_cleaned[pd.to_numeric(df_cleaned['ZIP CODE'], errors='coerce').isna()]['ZIP CODE'].unique())

# Replace any non-numeric values with pd.NA
df_cleaned['ZIP CODE'] = pd.to_numeric(df_cleaned['ZIP CODE'], errors='coerce')

# Now drop these as they are not valid ZIP codes
df_cleaned.dropna(subset=['ZIP CODE'], inplace=True)

# Try converting 'ZIP CODE' to 'Int64' again
df_cleaned['ZIP CODE'] = df_cleaned['ZIP CODE'].astype('Int64')

# Display the updated DataFrame shape and data type of 'ZIP CODE'
print("Updated DataFrame shape:", df_cleaned.shape)
print("Data type of 'ZIP CODE':", df_cleaned['ZIP CODE'].dtype)

df_cleaned['ZIP CODE'].head()



Unique non-numeric or placeholder values in 'ZIP CODE': ['     ']
Updated DataFrame shape: (893166, 22)
Data type of 'ZIP CODE': Int64


0    10017
1    11413
2    11434
3    10463
4    10301
Name: ZIP CODE, dtype: Int64

In [10]:
#Finding out the start and end date of the remaining data
min_date = df_cleaned['CRASH DATE'].min()
max_date = df_cleaned['CRASH DATE'].max()

print("Minimum date:", min_date)
print("Maximum date:", max_date)

Minimum date: 01/01/2013
Maximum date: 12/31/2023


I have now removed the null values and un-necessary columns from the data. I will proceed with creating the dimensions for my DW. Starting with the Contributing Factor Dimension:

In [11]:
# Extract unique values from both columns
unique_factors_1 = df_cleaned['CONTRIBUTING FACTOR VEHICLE 1'].dropna().unique()
unique_factors_2 = df_cleaned['CONTRIBUTING FACTOR VEHICLE 2'].dropna().unique()

# Combine and deduplicate the unique values from both columns
all_unique_factors = pd.Series(list(set(unique_factors_1) | set(unique_factors_2)))

# Create a DataFrame with IDs and descriptions
dim_contributing_factor = pd.DataFrame(all_unique_factors, columns=['contributing-factor-description'])

# Adding an ID column assuming a simple sequential ID is acceptable
dim_contributing_factor['contributing-factor-id'] = range(1, len(dim_contributing_factor) + 1)

# Display the DataFrame
print(dim_contributing_factor)
dim_contributing_factor.info()



                      contributing-factor-description  contributing-factor-id
0                                 Passing Too Closely                       1
1                               Windshield Inadequate                       2
2   Pedestrian/Bicyclist/Other Pedestrian Error/Co...                       3
3                    Lane Marking Improper/Inadequate                       4
4                              Other Lighting Defects                       5
5                          Listening/Using Headphones                       6
6                                   Vehicle Vandalism                       7
7         Traffic Control Device Improper/Non-Working                       8
8                             Prescription Medication                       9
9                                  Lost Consciousness                      10
10                              Accelerator Defective                      11
11                                                 80           

Next, I will create the vehicle Dimension


In [12]:
# Extract unique values from both vehicle type columns
unique_vehicle_types_1 = df_cleaned['VEHICLE TYPE CODE 1'].unique()
unique_vehicle_types_2 = df_cleaned['VEHICLE TYPE CODE 2'].unique()

# Combine and deduplicate the unique values from both columns
all_unique_vehicle_types = pd.Series(list(set(unique_vehicle_types_1) | set(unique_vehicle_types_2)))

# Create a DataFrame with IDs and descriptions
vehicle_type_df = pd.DataFrame(all_unique_vehicle_types, columns=['vehicle-type-description'])

# Adding an ID column assuming a simple sequential ID is acceptable
vehicle_type_df['vehicle-id'] = range(1, len(vehicle_type_df) + 1)
dim_vehicle = vehicle_type_df

# Display the DataFrame to verify
print(dim_vehicle)
dim_vehicle.info()

           vehicle-type-description  vehicle-id
0                             inter           1
1     SPORT UTILITY / STATION WAGON           2
2                           utility           3
3                             GATOR           4
4                             Small           5
...                             ...         ...
1299                          EBIKE        1300
1300                          bulld        1301
1301                        Go kart        1302
1302                       DELIVERY        1303
1303                           f550        1304

[1304 rows x 2 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1304 entries, 0 to 1303
Data columns (total 2 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   vehicle-type-description  1304 non-null   object
 1   vehicle-id                1304 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 20.5+ KB


Now, we create the Location Dimension. For this dimension, I tried adding an attribute called neighbourhood which I intended to infer from the location co-ordinates. However, the code to retrieve the neighbourhood of each row took too long to run. ChatGPT estimated it would take approximately 10.3 days to complete.

In [13]:
# Select necessary columns and ensure no rows with missing latitude or longitude
df_location = df_cleaned[['BOROUGH', 'ZIP CODE', 'LATITUDE', 'LONGITUDE', 'LOCATION', 'ON STREET NAME', 'CROSS STREET NAME']].dropna(subset=['LATITUDE', 'LONGITUDE'])

# Rename 'ZIP CODE' to 'zipcode' before creating the unique_zip_codes DataFrame
df_location.rename(columns={'ZIP CODE': 'zipcode'}, inplace=True)

# Extract unique ZIP codes and map them to a new ID column
unique_zip_codes = pd.DataFrame(df_location['zipcode'].unique(), columns=['zipcode'])
unique_zip_codes['location_ID'] = range(1, len(unique_zip_codes) + 1)

# Merge unique ZIP codes back to df_location to assign location_ID based on 'zipcode'
dim_location = pd.merge(df_location, unique_zip_codes, how='left', on='zipcode')

# Optionally combine LATITUDE and LONGITUDE into 'location_coordinates' if needed
dim_location['location-coordinates'] = dim_location['LATITUDE'].astype(str) + ', ' + dim_location['LONGITUDE'].astype(str)

# Rename columns to match SQL table definition
dim_location.rename(columns={
    'BOROUGH': 'borough',
    'LATITUDE': 'latitude',
    'LONGITUDE': 'longitude',
    'LOCATION': 'location_coordinates',
    'ON STREET NAME': 'on-street-name',
    'CROSS STREET NAME': 'cross-street-name'
}, inplace=True)

# Display the DataFrame to verify
print(dim_location)
dim_location.info()



              borough  zipcode   latitude  longitude     location_coordinates  \
0           MANHATTAN    10017  40.751440 -73.973970    (40.75144, -73.97397)   
1              QUEENS    11413  40.675884 -73.755770   (40.675884, -73.75577)   
2              QUEENS    11434  40.666840 -73.789410    (40.66684, -73.78941)   
3               BRONX    10463  40.872620 -73.904686   (40.87262, -73.904686)   
4       STATEN ISLAND    10301  40.631650 -74.087620    (40.63165, -74.08762)   
...               ...      ...        ...        ...                      ...   
893161       BROOKLYN    11215  40.672066 -73.990524  (40.672066, -73.990524)   
893162         QUEENS    11373  40.731968 -73.884780   (40.731968, -73.88478)   
893163  STATEN ISLAND    10306  40.570465 -74.109770   (40.570465, -74.10977)   
893164         QUEENS    11366  40.720955 -73.809350   (40.720955, -73.80935)   
893165       BROOKLYN    11203  40.655514 -73.927864  (40.655514, -73.927864)   

               on-street-na

Finally, we create the date dimension.

In [14]:
def time_of_day_bucket(hour):
    """ Categorize time into buckets based on hour."""
    if 0 <= hour < 4:
        return 'Late Night'
    elif 4 <= hour < 8:
        return 'Early Morning'
    elif 8 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 16:
        return 'Afternoon'
    elif 16 <= hour < 20:
        return 'Evening'
    else:
        return 'Night'

# Convert and create necessary time columns
df_cleaned['crash-time-stamp'] = pd.to_datetime(df_cleaned['CRASH DATE'].astype(str) + ' ' + df_cleaned['CRASH TIME'].astype(str))
df_cleaned['CRASH DATE'] = pd.to_datetime(df_cleaned['CRASH DATE'])

# Creating 'date-id' in df_cleaned
df_cleaned['date-id'] = df_cleaned['crash-time-stamp'].dt.strftime('%Y%m%d%H%M')

# Extract unique date-ids and associated rows
unique_dates = df_cleaned.drop_duplicates(subset='date-id')

# Initialize date_dimension using unique_dates
date_dimension = pd.DataFrame({
    'date-id': unique_dates['date-id'],
    'crash-date': unique_dates['CRASH DATE'],
    'crash-time': unique_dates['crash-time-stamp'],
    'hour': unique_dates['crash-time-stamp'].dt.hour,
    'am/pm-flag': unique_dates['crash-time-stamp'].dt.strftime('%p'),
    'time-of-day-bucket': unique_dates['crash-time-stamp'].dt.hour.apply(time_of_day_bucket),
    'date-iso-format': unique_dates['CRASH DATE'].apply(lambda x: x.isoformat()),
    'year-number': unique_dates['CRASH DATE'].dt.year,
    'quarter': unique_dates['CRASH DATE'].dt.quarter,
    'month-number': unique_dates['CRASH DATE'].dt.month,
    'day-number': unique_dates['CRASH DATE'].dt.day,
    'month-name': unique_dates['CRASH DATE'].dt.strftime('%B'),
    'day-name': unique_dates['CRASH DATE'].dt.strftime('%A'),
    'week-of-the-year': unique_dates['CRASH DATE'].dt.isocalendar().week,
    'week-of-the-month': unique_dates['CRASH DATE'].apply(week_of_month)
})

# Display the date_dimension DataFrame
print(date_dimension)
date_dimension.info()

             date-id crash-date          crash-time  hour am/pm-flag  \
0       202112141458 2021-12-14 2021-12-14 14:58:00    14         PM   
1       202112141650 2021-12-14 2021-12-14 16:50:00    16         PM   
2       202112142310 2021-12-14 2021-12-14 23:10:00    23         PM   
3       202112111943 2021-12-11 2021-12-11 19:43:00    19         PM   
4       202112131740 2021-12-13 2021-12-13 17:40:00    17         PM   
...              ...        ...                 ...   ...        ...   
893199  202404190727 2024-04-19 2024-04-19 07:27:00     7         AM   
893200  202404080840 2024-04-08 2024-04-08 08:40:00     8         AM   
893202  202404181425 2024-04-18 2024-04-18 14:25:00    14         PM   
893203  202404181515 2024-04-18 2024-04-18 15:15:00    15         PM   
893204  202404180640 2024-04-18 2024-04-18 06:40:00     6         AM   

       time-of-day-bucket      date-iso-format  year-number  quarter  \
0               Afternoon  2021-12-14T00:00:00         2021    

Now, we will create the Fact table

In [17]:
# Create mapping dictionaries from the dimension DataFrames
factor_id_map = dim_contributing_factor.set_index('contributing-factor-description')['contributing-factor-id'].to_dict()
vehicle_type_map = dim_vehicle.set_index('vehicle-type-description')['vehicle-id'].to_dict()
location_id_map = dim_location.set_index('zipcode')['location_ID'].to_dict()
date_type_map = date_dimension.set_index('crash-time')['date-id'].to_dict()


# Map contributing factor descriptions to IDs in df_cleaned
df_cleaned['contributing-factor-id'] = df_cleaned['CONTRIBUTING FACTOR VEHICLE 1'].map(factor_id_map)

# Map vehicle type descriptions to IDs in df_cleaned
df_cleaned['vehicle-id'] = df_cleaned['VEHICLE TYPE CODE 1'].map(vehicle_type_map)

# Map ZIP CODE to location IDs in df_cleaned
df_cleaned['location-id'] = df_cleaned['ZIP CODE'].map(location_id_map)

# Map crash-time-stamp to Date IDs in df_cleaned
df_cleaned['date-id'] = df_cleaned['crash-time-stamp'].map(date_type_map)

# Optionally, check how many IDs were successfully mapped and view the first few rows to confirm
print("Number of mapped contributing factors:", df_cleaned['contributing-factor-id'].notna().sum())
print("Number of mapped vehicle types:", df_cleaned['vehicle-id'].notna().sum())
print("Number of mapped locations:", df_cleaned['location-id'].notna().sum())
print("Number of mapped dates:", df_cleaned['date-id'].notna().sum())

print(df_cleaned[['CONTRIBUTING FACTOR VEHICLE 1', 'contributing-factor-id', 'VEHICLE TYPE CODE 1', 'vehicle-id', 'ZIP CODE', 'location-id','crash-time-stamp','date-id']].head())

Number of mapped contributing factors: 893166
Number of mapped vehicle types: 893166
Number of mapped locations: 893166
Number of mapped dates: 893166
    CONTRIBUTING FACTOR VEHICLE 1  contributing-factor-id  \
0             Passing Too Closely                       1   
1              Turning Improperly                      26   
2  Reaction to Uninvolved Vehicle                      38   
3                     Unspecified                      25   
4                     Unspecified                      25   

                   VEHICLE TYPE CODE 1  vehicle-id  ZIP CODE  location-id  \
0                                Sedan         241     10017            1   
1                                Sedan         241     11413            2   
2                                Sedan         241     11434            3   
3  Station Wagon/Sport Utility Vehicle         614     10463            4   
4                                Sedan         241     10301            5   

     crash-time-sta

In [18]:
df_cleaned.head()
df_cleaned.shape

(893166, 27)

In [19]:
# Initialize fact_table with the primary key column
fact_table = pd.DataFrame(df_cleaned['COLLISION_ID'].copy())
fact_table.columns = ['collision_id']

# Add number of persons killed
fact_table['number-of-persons-killed'] = df_cleaned['NUMBER OF PERSONS KILLED'].copy()

# Add number of pedestrians killed
fact_table['number-of-pedestrians-killed'] = df_cleaned['NUMBER OF PEDESTRIANS KILLED'].copy()

# Add number of cyclists killed
fact_table['number-of-cyclists-killed'] = df_cleaned['NUMBER OF CYCLIST KILLED'].copy()

# Add number of motorists killed
fact_table['number-of-motorists-killed'] = df_cleaned['NUMBER OF MOTORIST KILLED'].copy()

# Add number of persons injured
fact_table['number-of-persons-injured'] = df_cleaned['NUMBER OF PERSONS INJURED'].copy()

# Add number of pedestrians injured
fact_table['number-of-pedestrians-injured'] = df_cleaned['NUMBER OF PEDESTRIANS INJURED'].copy()

# Add number of cyclists injured
fact_table['number-of-cyclists-injured'] = df_cleaned['NUMBER OF CYCLIST INJURED'].copy()

# Add number of motorists injured
fact_table['number-of-motorists-injured'] = df_cleaned['NUMBER OF MOTORIST INJURED'].copy()

# Add vehicle-type-id (assuming it's correctly named and present)
fact_table['vehicle-id'] = df_cleaned['vehicle-id'].copy()

# Add location-id (if available and correctly named)
fact_table['location-id'] = df_cleaned['location-id'].copy()

# Add date-id
fact_table['date-id'] = df_cleaned['date-id'].copy()

# Add contributing-factor-id
fact_table['contributing-factor-id'] = df_cleaned['contributing-factor-id'].copy()

fact_table.head()

collision_id  number-of-persons-killed  number-of-pedestrians-killed  \
0       4486519                       0.0                             0   
1       4487127                       0.0                             0   
2       4486635                       0.0                             0   
3       4487040                       0.0                             0   
4       4487001                       0.0                             0   

   number-of-cyclists-killed  number-of-motorists-killed  \
0                          0                           0   
1                          0                           0   
2                          0                           0   
3                          0                           0   
4                          0                           0   

   number-of-persons-injured  number-of-pedestrians-injured  \
0                        0.0                              0   
1                        0.0                              0   
2                        2.0                              0   
3                        1.0                              0   
4                        1.0                              0   

   number-of-cyclists-injured  number-of-motorists-injured  vehicle-id  \
0                           0                            0         241   
1                           0                            0         241   
2                           0                            2         241   
3                           0                            1         614   
4                           0                            1         241   

   location-id       date-id  contributing-factor-id  
0            1  202112141458                       1  
1            2  202112141650                      26  
2            3  202112142310                      38  
3            4  202112111943                      25  
4            5  202112131740                      25

The fact table has been created successfully. Looks like I need to change some column types to integers before I begin pushing the data into the Data Warehouse.

In [20]:
# Convert columns to integer data type
fact_table['number-of-persons-killed'] = fact_table['number-of-persons-killed'].astype(int)
fact_table['number-of-persons-injured'] = fact_table['number-of-persons-injured'].astype(int)

fact_table.head()

fact_table.shape

(893166, 13)

Now, I will begin pushing my tables to the DW. First we begin by creating the connection to the DW

In [21]:
# Database connection URL
# Replace the placeholders with your actual database credentials
pwd = 'Land4you!'
database_url = f'postgresql://Salman:{pwd}@cis9440-hwdw-kss.postgres.database.azure.com/postgres'

# Create a SQLAlchemy engine
engine = create_engine(database_url)

Adding the data for my four dimensions, followed by the fact table

In [ ]:
date_dimension.to_sql('date_dimension', con=engine, if_exists='append', index=False)
dim_vehicle.to_sql('dim_vehicle', con=engine, if_exists='append', index=False)
dim_contributing_factor.to_sql('dim_contributing_factor', con=engine, if_exists='append', index=False)
dim_location.to_sql('dim_location', con=engine, if_exists='append', index=False)


166

In [ ]:
date_dimension.to_sql('date_dimension', con=engine, if_exists='append', index=False)


312

In [22]:
fact_table.to_sql('fact_table', con=engine, if_exists='append', index=False)

166

The data went to tables under the public tab of DataGrip

In [23]:
date_dimension.to_csv("date_dimension.csv",index=False)
dim_location.to_csv("dim_location.csv",index=False)
dim_contributing_factor.to_csv("dim_contributing_factor",index=False)
dim_vehicle.to_csv("dim_vehicle.csv",index=False)
fact_table.to_csv("fact_table.csv",index=False)

KeyboardInterrupt: 

In [24]:
# Generic Functions
def create_string(length):
    if isinstance(length, int) and length > 0:
        result_string = "(" + "?," * (length - 1) + "?)"
        return result_string

def insert_data(table_name, df):
    conn = pyodbc.connect(connection_string)
    cursor = conn.cursor()
    result = create_string(len(df.columns))
    # Insert data into the table
    insert_query = f"INSERT INTO {table_name} VALUES {result}"
    print(insert_query)
    cursor.executemany(insert_query, df.values.tolist())
    conn.commit()
    conn.close()